docs/source/ddp_comm_hooks.rst

```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
+
+DDP communication hook is a generic interface to control how to communicate
+gradients across workers.
+A few built-in communication hooks are provided,
+and users can easily apply any of these hooks to optimize communication.
+Besides, the hook interface can also support user-defined communication
+strategies for more advanced use cases.
+
+.. warning ::
+    DDP communication hooks can only support single process single device mode
+    on NCCL backend.
+
+How to Use A Communication Hook?
+--------------------------------
+
+To use a communication hook, the user just needs to let the DDP model register
+the hook before the training loop.
+
+.. automethod:: torch.nn.parallel.DistributedDataParallel.register_comm_hook
+
+Default Communication Hooks
+---------------------------
+
+Default communication hooks are simple **stateless** hooks, so the input state
+in ``register_comm_hook`` is either a process group or ``None``.
+
+.. automodule:: torch.distributed.algorithms.ddp_comm_hooks.default_hooks
+    :members:
```

Comment:
Just wanted to double check this doesn't pull in the private methods from `default_hooks` right? Also, we probably should not pull in `allreduce_fut` which is a public method (should it be private)?

Reply:
`_ allreduce_fut` is a private method, and only `allreduce_hook` and `fp16_compress_hook` are exposed.